File: ui-ngx/src/app/core/http/role.service.ts
///
/// Copyright © 2016-2026 The Thingsboard Authors
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.
///

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { PageLink } from '@shared/models/page/page-link';
import { PageData } from '@shared/models/page/page-data';
import { Role, RolePermission, Resource, Operation } from '@shared/models/role.models';
import { defaultHttpOptionsFromConfig, RequestConfig } from './http-utils';

@Injectable({
  providedIn: 'root'
})
export class RoleService {

  constructor(
    private http: HttpClient
  ) { }

  public getRoles(pageLink: PageLink, config?: RequestConfig): Observable<PageData<Role>> {
    return this.http.get<PageData<Role>>(`/api/role${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getRole(roleId: string, config?: RequestConfig): Observable<Role> {
    return this.http.get<Role>(`/api/role/${roleId}`, defaultHttpOptionsFromConfig(config));
  }

  public saveRole(role: Role, config?: RequestConfig): Observable<Role> {
    return this.http.post<Role>('/api/role', role, defaultHttpOptionsFromConfig(config));
  }

  public deleteRole(roleId: string, config?: RequestConfig): Observable<void> {
    return this.http.delete<void>(`/api/role/${roleId}`, defaultHttpOptionsFromConfig(config));
  }

  public getPermissions(roleId: string, config?: RequestConfig): Observable<RolePermission[]> {
    return this.http.get<RolePermission[]>(`/api/role/${roleId}/permissions`,
      defaultHttpOptionsFromConfig(config));
  }

  public updatePermissions(roleId: string, permissions: RolePermission[],
                          config?: RequestConfig): Observable<void> {
    return this.http.put<void>(`/api/role/${roleId}/permissions`, permissions,
      defaultHttpOptionsFromConfig(config));
  }

  public addPermissions(roleId: string, permissions: RolePermission[],
                       config?: RequestConfig): Observable<void> {
    return this.http.post<void>(`/api/role/${roleId}/permissions`, permissions,
      defaultHttpOptionsFromConfig(config));
  }

  public removePermissions(roleId: string, permissions: RolePermission[],
                          config?: RequestConfig): Observable<void> {
    return this.http.delete<void>(`/api/role/${roleId}/permissions`,
      { ...defaultHttpOptionsFromConfig(config), body: permissions });
  }

  public getAvailableResources(config?: RequestConfig): Observable<Resource[]> {
    return this.http.get<Resource[]>('/api/role/resources', defaultHttpOptionsFromConfig(config));
  }

  public getAvailableOperations(config?: RequestConfig): Observable<Operation[]> {
    return this.http.get<Operation[]>('/api/role/operations', defaultHttpOptionsFromConfig(config));
  }
}

File: common/dao-api/src/main/java/org/thingsboard/server/dao/role/RoleService.java
/**
 * Copyright © 2016-2026 The Thingsboard Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.thingsboard.server.dao.role;

import com.google.common.util.concurrent.ListenableFuture;
import org.thingsboard.server.common.data.id.RoleId;
import org.thingsboard.server.common.data.id.TenantId;
import org.thingsboard.server.common.data.page.PageData;
import org.thingsboard.server.common.data.page.PageLink;
import org.thingsboard.server.common.data.security.Role;
import org.thingsboard.server.common.data.security.RolePermission;
import org.thingsboard.server.dao.entity.EntityDaoService;

import java.util.List;
import java.util.Set;

public interface RoleService extends EntityDaoService {

    Role findRoleById(TenantId tenantId, RoleId roleId);

    ListenableFuture<Role> findRoleByIdAsync(TenantId tenantId, RoleId roleId);

    Role findRoleByTenantIdAndName(TenantId tenantId, String name);

    Role saveRole(Role role);

    Role saveRole(Role role, boolean doValidate);

    void deleteRole(TenantId tenantId, RoleId roleId);

    PageData<Role> findRolesByTenantId(TenantId tenantId, PageLink pageLink);

    List<Role> findSystemRoles();

    List<Role> findRolesByTenantIdAndIds(TenantId tenantId, List<RoleId> roleIds);

    Set<RolePermission> getRolePermissions(RoleId roleId);

    void updateRolePermissions(RoleId roleId, Set<RolePermission> permissions);

    void addRolePermissions(RoleId roleId, Set<RolePermission> permissions);

    void removeRolePermissions(RoleId roleId, Set<RolePermission> permissions);

    void createDefaultTenantRoles(TenantId tenantId);

    long countUsersByRoleId(RoleId roleId);

    void deleteRolesByTenantId(TenantId tenantId);
}
